python/google/protobuf/pyext/descriptor.cc

```diff
@@ -99,43 +136,63 @@ bool _CalledFromGeneratedFile(int stacklevel) {
   if (frame == nullptr) {
     return false;
   }
+  Py_INCREF(frame);
   while (stacklevel-- > 0) {
-    frame = frame->f_back;
+    PyFrameObject* next_frame = PyFrame_GetBack(frame);
+    Py_DECREF(frame);
+    frame = next_frame;
     if (frame == nullptr) {
       return false;
     }
   }
 
-  if (frame->f_code->co_filename == nullptr) {
+  PyCodeObject* frame_code = PyFrame_GetCode(frame);
+  if (frame_code->co_filename == nullptr) {
+    Py_DECREF(frame_code);
+    Py_DECREF(frame);
     return false;
   }
   char* filename;
   Py_ssize_t filename_size;
-  if (PyString_AsStringAndSize(frame->f_code->co_filename,
+  if (PyString_AsStringAndSize(frame_code->co_filename,
                                &filename, &filename_size) < 0) {
     // filename is not a string.
+    Py_DECREF(frame_code);
+    Py_DECREF(frame);
     PyErr_Clear();
     return false;
   }
+  Py_DECREF(frame_code);
   if ((filename_size < 3) ||
       (strcmp(&filename[filename_size - 3], ".py") != 0)) {
     // Cython's stack does not have .py file name and is not at global module
     // scope.
+    Py_DECREF(frame);
     return true;
   }
   if (filename_size < 7) {
     // filename is too short.
+    Py_DECREF(frame);
     return false;
   }
   if (strcmp(&filename[filename_size - 7], "_pb2.py") != 0) {
     // Filename is not ending with _pb2.
+    Py_DECREF(frame);
     return false;
   }
 
-  if (frame->f_globals != frame->f_locals) {
+  PyObject* frame_globals = PyFrame_GetGlobals(frame);
+  PyObject* frame_locals = PyFrame_GetLocals(frame);
+  if (frame_globals != frame_locals) {
     // Not at global module scope
+    Py_DECREF(frame_globals);
+    Py_XDECREF(frame_locals);
+    Py_DECREF(frame);
     return false;
   }
+  Py_DECREF(frame_globals);
+  Py_XDECREF(frame_locals);
+  Py_DECREF(frame);
```

Comment:
rather than an ever increasing number of decrefs, please add an error exist that you jump to with a goto that handles all the decrefs

Reply:
Add exit handler